### ISSUE TITLE
Sometimes ads will get through on YouTube on Playlist

### ISSUE THREAD

**Issue by bsclifton** [brave/brave-ios#3457]
### Description: 
Definitely need some help characterizing this

I believe the problem is that ads can leak through to YouTube. If you visit the video, it gets cached by browser, and then you visit again and add to playlist... our player will play the ads because of a race condition where the rules aren't loaded in time (since it loaded from cache)

Known conditions:
- user must be logged into their Google/YouTube account
- video must be a monetized one (ex: https://youtu.be/PvYhZT99g1s)

Is there anything we can do to prevent this from happening? I know our player will be hitting the stream directly (unless it's downloaded). Is the stream redirecting to an ad?

**Comment by bsclifton** [comment]
Fix for initial report here:
https://github.com/brave/brave-ios/commit/52fcb972566639c4c1e12f35a136d17cf8149643

No additional reports that I know of since then

**Comment by Brandon-T** [comment]
The url for videos with ads always ends in `pltype=adhost` and for regular videos, it ends in `pltype=content`.
So to detect if an ad is playing, I check now if the query parameter `pltype` has the value `adhost`.

One way to reproduce the issue:

1. Go to youtube and login!
2. Go to monetized video and hold the finger on the refresh button so that it brings up a menu.
3. In that menu, "Request Desktop Site" as the issue does NOT happen on mobile sites!
4. Add the desktop video to playlist.
5. Open playlist and play the item. You can play the item by tapping on it. If you tap on it multiple times, sometimes it will fetch an ad.

**Comment by Brandon-T** [comment]
Current Fix: https://github.com/brave/brave-ios/blob/development/Client/Frontend/Browser/Playlist/Cache/PlaylistCacheLoader.swift#L552

1. Force adblock to be enabled by creating a non-persistent CoreData domain with all values for adblock set to true.. This means we have to ignore the user's preferences in settings for the invisible webView:
```swift
// Identify specific block lists that need to be applied to the requesting domain
let domainForShields = Domain.getOrCreate(forUrl: mainDocumentURL, persistent: false)
                
// Force adblocking on
domainForShields.shield_allOff = 1
domainForShields.shield_adblockAndTp = true
domainForShields.shield_httpse = true
                
let (on, off) = BlocklistName.blocklists(forDomain: domainForShields)
let controller = webView.configuration.userContentController
```

2. We removed the DownloadHelper for invisible webView because every time a page is visited, it'd attempt to download but fail when the mime-type is `video/mp4`.. but when it is `video/html` it skips the downloader.. and not all websites have the correct mime-types for media items.

3. To get this all working, we duplicated Browser code from lines 552 to 772 and modified it heavily to always have ads on, to inject content blockers, ignore downloading and mime-types and to disable all unnecessary callbacks.

4. Invisible WebView isn't running anything until it is added to the window. We make the webView visible for a split second and then remove it from the screen so quickly that the user doesn't see it.. This way, all scripts and page load is triggered and so is content blockers. Our webView is an invisible "Tab"! It is not just a `WKWebView`.
